### PR TITLE
Enable recommended golang-ci linters

### PR DESCRIPTION
**Pull request by AndrewSirenko**
**Is this a bug fix or adding new feature?**
Refactor

**What is this PR about? / Why do we need it?**
We rely on golang-ci's suite of linters to analyze our code for potential errors or style issues. We have had most linters disabled for a while, instead relying on PR reviewers to catch common errors. 

This PR re-enables golang-ci's linters, and refactors the code base to pass those linters.

Please see .golangci.yml for a list of enabled/disabled linters and individual commit messages for the code changes required to enable each of them. 

It was requested that this be a single PR, though if it takes too long to review I will split it into two PRs: 
- one for zero-risk changes (style)
- and one for slightly risky changes (refactoring) 

**What testing is done?** 
Manual testing, CI, make verify

### CODE REVIEW

File: cmd/hooks/prestop.go
Comment (AndrewSirenko) [inline cmd/hooks/prestop.go:63]: Changing fmt.Errorf -> errors.New if Errorf is not necessary is one of the many performance improvements  `perfsprint` linter suggests.

I'm open to ignoring this specific rule if it is too annoying to future maintainers, but it's not too big of a deal ¯\_(ツ)_/¯

File: .golangci.yml
Comment (AndrewSirenko) [inline .golangci.yml:142]: This linter is a 2-3 point task on its own to add meaningful error wraps for external package errors.

File: .golangci.yml
Comment (AndrewSirenko) [inline .golangci.yml:146]: Took a stab at this but there are many tests that would require nolint because they are not written to be parallelized.

File: .golangci.yml
Comment (AndrewSirenko) [inline .golangci.yml:144]: Many many globals, 2+ point risky task on its own.

File: .golangci.yml
Comment (ConnorJC3) [inline .golangci.yml:39]: Probably better to use `enable-all` (and the list of disables below) rather than listing every enable _and_ disable?

File: .golangci.yml
Comment (ConnorJC3) [inline .golangci.yml:137]: This linter is missing a comment why we're disabling it. Also, isn't this a pretty bad anti-pattern - maybe it should be in the "Consider adding" section below?

File: cmd/main.go
Comment (ConnorJC3) [inline cmd/main.go:110]: This one we should probably put back in and `nolint`, it's normal to print version info entirely by itself, it shouldn't come with a timestamp like a log entry.

File: .golangci.yml
Comment (AndrewSirenko) [reply .golangci.yml:39]: I thought about this but then every time golang-ci updates we risk blocking CI because of new linters being added. 

Also there are some warnings because deprecated linters are still enabled in enable-all. 

IMO better to explicitly opt in to avoid this annoyance. I see no significant drawback to being more explicit about the linters we allow, we can always add more linters later.

File: .golangci.yml
Comment (ConnorJC3) [reply .golangci.yml:137]: Given that it's used in a bunch of places I don't think it would be reasonable at this point to enable it in this PR (unless you really want to volunteer to do the extra work). but I would suggest we move it down to the "maybe enable later" section so that we can make that decision in the future.

File: .golangci.yml
Comment (AndrewSirenko) [reply .golangci.yml:39]: > The primary drawback is that we have to search for and explicitly enable any new linter

Yep totally agree that we will miss out on new linters until someone explicitly adds them. (Though is that a bad thing)

I thought about adding a dependency upgrades runbook step of "If upgrading golang-ci, run with `--enable-all` to spot any new linters" but that is even more annoying than upgrading + manually disabling...

> deprecated linters is unfortunate

Yeah I couldn't find anything either. I'll consider raising an issue? But to be fair deprecated linter warnings are a small annoyance. 

---

Thanks for your thoughts, I don't have a strong opinion either way, it's a two way door that boils down to: 

"Do we want the person doing dependency upgrades to enable/disable new linters as they are added, or should we batch those decisions every once in a while"

I'll default to enabling all linters as you originally suggested unless @ElijahQuinones or @torredil have a preference on ^^

File: pkg/driver/controller.go
Comment (torredil) [inline pkg/driver/controller.go:59]: Could we remove `trueStr` and simplify this to 
```
const isManagedByDriver = "true"
```

File: pkg/cloud/volume_limits.go
Comment (torredil) [inline pkg/cloud/volume_limits.go:31]: Unless its already in the backlog, could we create an issue about this so as to not forget it?

File: cmd/main.go
Comment (torredil) [inline cmd/main.go:111]: It's not clear why this comment is added unless you read through the PR discussion. Suggest adding that clarity directly in the comment so that future maintainers don't accidentally change this.

**Review by torredil — COMMENTED** [review]
Largely lgtm 👍 
Left a few non-blocking comments.

/approve

File: pkg/driver/controller.go
Comment (ElijahQuinones) [inline pkg/driver/controller.go:488]: If we are changing cap to capability should we change caps to capabilities ?

File: pkg/driver/controller.go
Comment (ConnorJC3) [reply pkg/driver/controller.go:59]: `trueStr` is used in other places (where "is managed by driver" doesn't make sense) so I don't think we can - if we want to remove one of the two it should be `isManagedByDriver`

File: pkg/driver/controller_modify_volume.go
Comment (torredil) [reply pkg/driver/controller_modify_volume.go:93]: https://staticcheck.dev/docs/checks#ST1005

Comment (ElijahQuinones) [reply pkg/driver/controller_modify_volume.go:93]: Thank you Eddie resolving comment !

File: pkg/driver/request_coalescing_test.go
Comment (AndrewSirenko) [reply pkg/driver/request_coalescing_test.go:222]: This is unrelated to this PR

I'm impartial either way for the constant name.

File: pkg/driver/controller.go
Comment (AndrewSirenko) [reply pkg/driver/controller.go:488]: I was changing cap to capability to avoid colliding with built-in `cap` function, which this linter spotted.

Open to changing controllerCaps to controllerCapabilities if others want that in this PR.

File: pkg/driver/controller.go
Comment (AndrewSirenko) [reply pkg/driver/controller.go:59]: I left `isManagedByDriver` in because it was used in quite a few places to enhance readability. Can replace if we want.

I do not think we should be using isManagedByDriver to signify "true" in other places in the file.

File: pkg/driver/controller.go
Comment (ConnorJC3) [reply pkg/driver/controller.go:488]: Honestly I think that `caps` is clear enough in the context (and especially in `controllerCaps` keeps the variable size within reason), but don't have a strong opinion either way.

Comment (AndrewSirenko) [reply pkg/driver/controller.go:488]: Sounds good, resolving then.

**Review by ConnorJC3 — COMMENTED** [review]
/lgtm (other than needing a squash - please don't approve until those fixup commits are squashed)

**Review by ConnorJC3 — APPROVED** [review]
/lgtm

**Pull request merged**
